Allow projects to override implicit PackageReference items with item Include syntax

dsplaisted opened the pull request:
This is mutually exclusive with #625.

This allows the implicit package references to be overridden with the `<PackageReference Include="..." Version="..." />` syntax, which makes for a more consistent experience for people who manually edit their project files, as opposed to `<PackageReference Update="..." Version="..." />`.

The `Update` syntax will still work and that's what will be written to the project file if you use the Package Manager UI to upgrade.

dsplaisted committed:
Allow projects to override implicit PackageReference items with item Include syntax

dsplaisted committed:
Add warning message for when projects override implicit PackageReference items

dsplaisted commented:
I've updated this PR to log a warning message when an implicit `PackageReference` is overridden.  Here's what it currently looks like:

> C:\git\dotnet-sdk\bin\Debug\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.DefaultItems.targets(64,5): warning : A PackageReference for 'NETStandard.Library' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://github.com/dotnet/sdk

We also need to create an fwLink to use for the "More Information" URL.  @terrajobst or @blackdwarf, can you do that?

blackdwarf commented:
@dsplaisted I can create an aka.ms link (same thing), but we cannot point people to the SDK repo, since that doesn't seem to be a very nice solution. We cannot expect people to read a bunch of target files to understand something. If we go down this way, we need to create a document somewhere. I've created https://aka.ms/sdkimplicitrefs that points to a currently non-existent doc file in dotnet/core. 

/cc @richlander

dsplaisted committed:
Update link for more information on implicit PackageReference items

dsplaisted commented:
@blackdwarf Yes, the SDK repo was just a placeholder, I was expecting that the link would point to a documentation page that's to be written.

We probably want a similar page and link for the message in PR #624.  Do you think we should have two separate links, one for information on implicit `PackageReference` and one for the other implicit items (`Compile`, `EmbeddedResource`, `Content`, and `None`), or just one link?  I would suggest two separate aka.ms links, and you can always point them at the same page later if that makes sense.

dsplaisted committed:
Move implicit PackageReference logic to DefaultItems.props and DefaultItems.targets

dsplaisted committed:
Deduplicate implicit package references in evaluation instead of a target

dsplaisted committed:
Add inference logic for RuntimeFrameworkVersion and implicit package reference for Microsoft.NETCore.App

dsplaisted commented:
I've updated this PR to contain what I believe are all the changes we've agreed on for PackageReference deduplication, warnings, and inference with the RuntimeFrameworkVersion and package version for Microsoft.NETCore.App.

Relevant properties:

- `DisableImplicitFrameworkReferences` - If set to `true`, disables implicit `PackageReference` items
- `NetStandardImplicitPackageVersion` - The version of `NETStandard.Library` to implicitly reference.  Defaults to `1.6.1`
- `NetCoreAppImplicitPackageVersion` - The version of `Microsoft.NETCore.App` to implicitly reference.  Defaults to `RuntimeFrameworkVersion` if that property is set, otherwise defaults to `1.1.0`
- `RuntimeFrameworkVersion` - The version of the runtime that the app should run on, and is what is written as the version to the runtimeconfig.json file.  Defaults to be the same as the `TargetFrameworkVersion`

This means that generally a project can just specify the `TargetFramework` and you will get the correct version of everything.  If you want to specify a specific patch version, then you can set the `RuntimeFrameworkVersion` in your project.

This change means that projects targeting netcoreapp1.0 will by default specify version `1.0.0` as the version of the runtime that they require in the runtimeconfig.json.  The host policy will use a later patch if it is installed, so if `1.0.1` or `1.0.3` is installed, it will use that version.  We could also set the framework version to the highest patch version that the SDK knows about, ie `1.0.3` right now if targeting netcoreapp1.0.  My inclination, however, is not to do this in the SDK.

@srivatsn @nguerrera @natidea @davidfowl @terrajobst for review

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props:
```diff
@@ -26,4 +26,17 @@ Copyright (c) .NET Foundation. All rights reserved.
+  </ItemGroup>
+
+  <!-- The reference to the Microsoft.NETCore.App currently needs to match the target framework version.  So mark it as
+       ReadOnly so it won't show updates to the package in the package manager UI. -->
```
`ReadOnly` comment is out of date.

dsplaisted replied:
Thanks for the catch, I've fixed it.

dsplaisted committed:
Remove TODO and outdated comment

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets:
```diff
@@ -52,6 +52,94 @@ Copyright (c) .NET Foundation. All rights reserved.
+      <ImplicitPackageReferenceInformationLink>https://aka.ms/sdkimplicitrefs</ImplicitPackageReferenceInformationLink>
+    </PropertyGroup>
+
+    <NETSdkError Condition="'@(_ImplicitPackageReference)' != '' And '@(_OverridingPackageReference)' != ''"
```
Should there be a way to turn this warning off?

eerhardt replied:
I guess `DisableImplicitFrameworkReferences` should turn it off, right?  No change needed?

---
In reply to: [96059620](https://github.com/dotnet/sdk/pull/633#discussion_r96059620) [](ancestors = 96059620)

dsplaisted replied:
@eerhardt Yes, `DisableImplicitFrameworkReferences` is what you would use

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets:
```diff
@@ -52,6 +52,92 @@ Copyright (c) .NET Foundation. All rights reserved.
+
+    <!-- Add the override item back -->
+    <PackageReference Include="@(_OverridingPackageReference)"
+                      Condition="'@(_ImplicitPackageReference)' != '' And '@(_OverridingPackageReference)' != ''" />
```
Is remove/readd required because Remove can't filter by metadata?

dsplaisted replied:
Yes, I wasn't able to figure out a way to Remove only the item with specific metadata.

rainersigwald replied:
I can't either. In a target, you'd batch to evaluate the condition multiple times but that wouldn't work here. We'll have to see if this comes up often enough to worry about providing convenience for.

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets:
```diff
@@ -52,6 +52,92 @@ Copyright (c) .NET Foundation. All rights reserved.
+      <ImplicitPackageReferenceInformationLink>https://aka.ms/sdkimplicitrefs</ImplicitPackageReferenceInformationLink>
+    </PropertyGroup>
+
+    <NETSdkError Condition="'@(_ImplicitPackageReference)' != '' And '@(_OverridingPackageReference)' != ''"
```
I've seen comparisons like this collapsed to a single `'@(_ImplicitPackageReference)@(_OverridingPackageReference)' != ''`. Just FYI, no reason to change this.

srivatsn commented:
Tagging @MattGertz for approval.

rainersigwald reviewed (approved):
Looks like a fine implementation.

srivatsn commented:
@dsplaisted can you fill out the template?

dsplaisted commented:
@srivatsn @MattGertz  Here's the escrow template:

#### Scenario

This covers various scenarios, including

- Developers who created projects with RC2 and update to RC3 or later.  This will generate a warning message letting them know they can remove the `PackageReference` that is now implicit from their project files
- Retargeting .NET Core projects.  You won't get in a broken state if you change the target framework version without updating a corresponding package version, or if you upgrade the Microsoft.NETCore.App package without changing your target framework version
- Targeting a specific patch version of the runtime - You can now do this by setting the `RuntimeFrameworkVersion` property

#### Bug

n/a

#### Workarounds

Depending on the scenario, there were various workarounds, which generally involved knowing what to do with your project file and what not to do.  Overall, this change serves to create a "pit of success" instead of a "pit of failure".

#### Risk

Low / Medium - Tests should cover any impacted scenarios

#### Performance Impact

Low - The changes are part of MSBuild's evaluation pass

#### Regression Analysis

n/a

terrajobst commented on src/Tasks/Microsoft.NET.Build.Tasks/NETSdkError.cs:
```diff
@@ -24,13 +24,22 @@ public class NETSdkError : TaskBase
         /// </summary>
         public string[] FormatArguments { get; set; }
 
+        public bool WarningOnly { get; set; }
```
I assume that's the default? Why do we need the ability to error out?

dsplaisted replied:
This is a generic task that is used to generate localized errors from the Sdk.  This is the first time we needed it to generate a warning instead of an error, so I added that capability.

nguerrera replied:
I would prefer NETSdkWarning and NETSdkError tasks just like the built-in Error and Warning tasks. It is weird to read Error then Warning only.

dsplaisted commented:
When I tried these changes together with Visual Studio, the implicit package references stopped working.  VS wasn't seeing them, leading to compile errors since none of the Framework APIs were being referenced.  This seems to be the same issue as https://github.com/Microsoft/msbuild/issues/1475, where an item `Remove` with a condition on it is always getting applied when VS does the evaluation.

So, we're going to go back to the implementation where the deduplication is done in a task, and we will make a corresponding change in the Roslyn project system to apply the same filtering.

dsplaisted committed:
Switch back to deduplicating implicit PackageReference items in a target instead of in evaluation

The evaluation method wasn't working correctly in Visual Studio, seemingly because of https://github.com/Microsoft/msbuild/issues/1475

dsplaisted committed:
Run PackageReference de-dupe target before ResolvePackageReferencesDesignTime so that VS can get de-duped list

srivatsn committed:
Merge branch 'dev15-rc3' into dedupe-packagereference

srivatsn committed:
Rename ResolvePackageReferencesDesignTime to CollectPackageReferences

srivatsn commented:
Approved by JoC over email.
